File: app/src/main/java/com/nextcloud/talk/webrtc/AudioRoutePolicy.java
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.webrtc;

import java.util.Set;

final class AudioRoutePolicy {
    private AudioRoutePolicy() {
    }

    static WebRtcAudioManager.AudioDevice selectAudioDevice(
            Set<WebRtcAudioManager.AudioDevice> availableDevices,
            WebRtcAudioManager.AudioDevice userSelectedDevice,
            WebRtcAudioManager.AudioDevice defaultDevice,
            boolean hasWiredHeadset,
            boolean bluetoothConnected) {
        if (bluetoothConnected) {
            return WebRtcAudioManager.AudioDevice.BLUETOOTH;
        }

        if (hasWiredHeadset) {
            return WebRtcAudioManager.AudioDevice.WIRED_HEADSET;
        }

        if (userSelectedDevice != WebRtcAudioManager.AudioDevice.NONE
                && userSelectedDevice != WebRtcAudioManager.AudioDevice.BLUETOOTH
                && availableDevices.contains(userSelectedDevice)) {
            return userSelectedDevice;
        }

        if (defaultDevice != WebRtcAudioManager.AudioDevice.NONE && availableDevices.contains(defaultDevice)) {
            return defaultDevice;
        }

        if (availableDevices.contains(WebRtcAudioManager.AudioDevice.EARPIECE)) {
            return WebRtcAudioManager.AudioDevice.EARPIECE;
        }
        if (availableDevices.contains(WebRtcAudioManager.AudioDevice.SPEAKER_PHONE)) {
            return WebRtcAudioManager.AudioDevice.SPEAKER_PHONE;
        }
        return WebRtcAudioManager.AudioDevice.NONE;
    }

    static boolean shouldPreferBluetooth(
            WebRtcAudioManager.AudioDevice userSelectedDevice,
            boolean bluetoothCurrentlyPreferred,
            boolean bluetoothExpected,
            boolean bluetoothUnavailable) {
        if (userSelectedDevice == WebRtcAudioManager.AudioDevice.BLUETOOTH) {
            return true;
        }
        if (userSelectedDevice != WebRtcAudioManager.AudioDevice.NONE) {
            return false;
        }
        if (bluetoothExpected) {
            return true;
        }
        return bluetoothCurrentlyPreferred && !bluetoothUnavailable;
    }
}

File: app/src/main/java/com/nextcloud/talk/webrtc/AudioRoutePolicy.kt
/*
 * Nextcloud Talk - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
package com.nextcloud.talk.webrtc

import android.media.AudioDeviceInfo
import com.nextcloud.talk.webrtc.WebRtcAudioManager.AudioDevice

internal object AudioRoutePolicy {
    @JvmStatic
    fun selectAudioDevice(
        availableDevices: Set<AudioDevice>,
        userSelectedDevice: AudioDevice,
        defaultDevice: AudioDevice,
        hasWiredHeadset: Boolean,
        bluetoothConnected: Boolean
    ): AudioDevice =
        when {
            bluetoothConnected -> AudioDevice.BLUETOOTH
            hasWiredHeadset -> AudioDevice.WIRED_HEADSET
            userSelectedDevice != AudioDevice.NONE &&
                userSelectedDevice != AudioDevice.BLUETOOTH &&
                availableDevices.contains(userSelectedDevice) -> userSelectedDevice
            defaultDevice != AudioDevice.NONE && availableDevices.contains(defaultDevice) -> defaultDevice
            availableDevices.contains(AudioDevice.EARPIECE) -> AudioDevice.EARPIECE
            availableDevices.contains(AudioDevice.SPEAKER_PHONE) -> AudioDevice.SPEAKER_PHONE
            else -> AudioDevice.NONE
        }

    @JvmStatic
    fun shouldPreferBluetooth(
        userSelectedDevice: AudioDevice,
        bluetoothCurrentlyPreferred: Boolean,
        bluetoothExpected: Boolean,
        bluetoothUnavailable: Boolean
    ): Boolean =
        when {
            userSelectedDevice == AudioDevice.BLUETOOTH -> true
            userSelectedDevice != AudioDevice.NONE -> false
            bluetoothExpected -> true
            else -> bluetoothCurrentlyPreferred && !bluetoothUnavailable
        }

    @JvmStatic
    fun shouldSetCommunicationDevice(currentRouteMatches: Boolean, routeSelectionMustBeReasserted: Boolean): Boolean =
        !currentRouteMatches || routeSelectionMustBeReasserted

    @JvmStatic
    fun shouldSelectBeforeBluetoothTeardown(bluetoothStopNeeded: Boolean, targetDevice: AudioDevice): Boolean =
        bluetoothStopNeeded && targetDevice != AudioDevice.NONE && targetDevice != AudioDevice.BLUETOOTH

    @JvmStatic
    fun canFinishBluetoothTeardown(
        bluetoothStopNeeded: Boolean,
        targetMustBeSelectedFirst: Boolean,
        targetSelectionSucceeded: Boolean
    ): Boolean = bluetoothStopNeeded && (!targetMustBeSelectedFirst || targetSelectionSucceeded)

    @JvmStatic
    fun isWiredCommunicationOutput(type: Int, isSink: Boolean): Boolean =
        when (type) {
            AudioDeviceInfo.TYPE_WIRED_HEADSET,
            AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
            AudioDeviceInfo.TYPE_USB_HEADSET,
            AudioDeviceInfo.TYPE_USB_DEVICE,
            AudioDeviceInfo.TYPE_USB_ACCESSORY -> isSink
            else -> false
        }
}
